src/shogun/clustering/KMeans.cpp

```diff
@@ -4,446 +4,189 @@
-	SGVector<int32_t>::range_fill_vector(temp, lhs_size, 0);
-	CMath::permute(temp);
+	int32_t lhs_size=lhs->get_num_vectors();
+	int32_t dim=lhs->get_num_features();
 
-	for (int32_t i=0; i<k; i++)
-	{
-		const int32_t cluster_center_i=temp[i];
-		SGVector<float64_t> vec=lhs->get_feature_vector(cluster_center_i);
+	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
+	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
 
-		for (int32_t j=0; j<dimensions; j++)
-			mus(j,i)=vec[j];
+	SGVector<int32_t> cluster_assignments=SGVector<int32_t>(lhs_size);
+	cluster_assignments.zero();
 
-		lhs->free_feature_vector(vec, cluster_center_i);
-	}
+	/* Weights : Number of points in each cluster */
+	SGVector<int32_t> weights_set(num_centers);
+	weights_set.zero();
+	/* Initially set all weights for zeroth cluster, Changes in assignement step */
+	weights_set[0]=lhs_size;
 
-	SG_UNREF(lhs);
+	distance->precompute_lhs();
 
-}
+	int32_t changed=1;
+	int32_t iter;
 
-void CKMeans::compute_cluster_variances()
-{
-	/* compute the ,,variances'' of the clusters */
-	for (int32_t i=0; i<k; i++)
+	for(iter=0; iter<max_iter; iter++)
 	{
-		float64_t rmin1=0;
-		float64_t rmin2=0;
-
-		bool first_round=true;
-
-		for (int32_t j=0; j<k; j++)
-		{
-			if (j!=i)
+		if (iter==max_iter-1)
+			SG_SWARNING("KMeans clustering has reached maximum number of ( %d ) iterations without having converged. \
+				   	Terminating. \n", iter)
+
+		changed=0;
+		rhs_mus->copy_feature_matrix(centers);
+
+		distance->precompute_rhs();
+
+#pragma omp parallel for firstprivate(lhs_size, dim, num_centers) \
+		shared(centers, cluster_assignments, weights_set) \
+		reduction(+:changed) if (!fixed_centers)
+		/* Assigment step : Assign each point to nearest cluster */
+		for (int32_t i=0; i<lhs_size; i++)
+		{ 
+			const int32_t cluster_assignments_i=cluster_assignments[i];
+			int32_t min_cluster, j;
+			float64_t min_dist, dist;
+			
+			min_cluster=0;
+		   	min_dist=distance->distance(i,0);
+			for (j=1; j<num_centers; j++)
 			{
-				int32_t l;
-				float64_t dist = 0;
-
-				for (l=0; l<dimensions; l++)
+				dist=distance->distance(i,j);
+				if (dist<min_dist)
 				{
-					dist+=CMath::sq(
-							mus.matrix[i*dimensions+l]
-									-mus.matrix[j*dimensions+l]);
+					min_dist=dist;
+					min_cluster=j;
 				}
+			}
 
-				if (first_round)
-				{
-					rmin1=dist;
-					rmin2=dist;
-					first_round=false;
-				}
-				else
+			if (min_cluster!=cluster_assignments_i)
+			{
+				changed++;
+#pragma omp atomic
+				weights_set[min_cluster]+= 1;
+#pragma omp atomic
+				weights_set[cluster_assignments_i]-= 1;
+
+				if(fixed_centers)
 				{
-					if ((dist<rmin2) && (dist>=rmin1))
-						rmin2=dist;
+					SGVector<float64_t>vec=lhs->get_feature_vector(i);
+					
+					/* mu_new = mu_old + (x - mu_old)/(w) */					
+					for (j=0; j<dim; j++)
+					{
+						centers(j, min_cluster)+=
+							(vec[j]-centers(j, min_cluster)) / weights_set[min_cluster];
```

Comment:
only thing: a `/` operator (division) is usually much more costly than multiplication...
so precomputing `x = 1/weights_set[min_cluster]` and then just do
```
vec[j]-x*centers(j, min_cluster))
```
of course it'd be good to check what does the compiler do when you do `-O3`